### DISCUSSION TITLE
jukebox satellites

### DISCUSSION THREAD

**Discussion by tcurdt** [navidrome/navidrome#3539] [Ideas]
I would like to propose a few changes to the jukebox mode (and I am also willing to work on them if deemed acceptable)

https://www.navidrome.org/docs/usage/jukebox/

Instead of playing on the server I want to play music on satellites. The main navidrome server would basically tell the satellites to play a the song. The satellites would reach out to navidrome to play the song.

Think of if as an open source chromecast replacement.

The most basic implementation could allow for URLs as targets

```
Jukebox.Devices = [
    [ "internal", "coreaudio/BuiltInSpeakerDevice" ],
    [ "satellite1", "https://satellite1.music.home" ],
    [ "satellite2", "https://satellite2.music.home" ]
]
```

The satellite could be either navidrome in a special satellite mode or a separate project.

#### Next level of the idea:

The jukebox does not have to be a single jukebox per server. In theory there could be a jukebox per user even. Yet, you might also want to be able to share a jukebox between users. But that means we would also need to define multiple instances of jukeboxes.

```
Jukeboxes = [
    [ "default", "internal" ],
    [ "satellites", "satellites*" ],
    [ "userA", "internal, satellites*" ]
]
```

**Anyone in support of that idea?**

**Comment by metalheim** [comment]
You can do something *like* this with configuring [snapcast ](https://github.com/badaix/snapcast/tree/develop) as an output
https://www.navidrome.org/docs/usage/jukebox/#the-mpvcmdtemplate--snapcast-integration

And then have a snapcast client running as "satellites". 
Clients can be turned on/off with snapcast and even be grouped to have time-synced multi-room audio.

It's not as smooth as having "navidrome-native" "thin clients (satellites)", but at least its available right now

**Reply by tcurdt** [reply]
Thanks for the pointer. That sure is interesting, too.
I'll see if I can this running as a first test.

**Reply by tcurdt** [reply]
Not sure I fully understand the jukebox setup yet

```
Jukebox.Devices = [
    # "symbolic name" used in the UI
    # "device" see mpv --audio-device=help
    # "symbolic name " "device"
    [ "snap", "alsa" ]
]

# MPVCmdTemplate = "mpv --no-audio-display --pause %f --input-ipc-server=%s --audio-channels=stereo --audio-samplerate=48000 --audio-format=s16 --ao=pcm --ao-pcm-file=/tmp/snapfifo"

# Using device "auto" if missing
Jukebox.Default = "snap"

Jukebox.Enabled = true
```

The idea is that the navidrome passes the file to mpv, which decodes the audio and passes the raw audio to the snapserver (via fifo) which then re-encodes it and sends it out the clients?

Does the "device" even matter in the above config?

**Reply by metalheim** [reply]
> Does the "device" even matter in the above config?

No. Device doesnt matter and doesnt need to be configured

**Reply by tcurdt** [reply]
Hm... seems like there a couple of people struggling to get this working:

https://github.com/navidrome/navidrome/issues/3022
https://github.com/navidrome/navidrome/issues/3256

From the docs and and reading through the issues it sounds this should work.

```
Jukebox.Devices = [
    [ "snap", "whatever" ]
]

MPVPath = ""
MPVCmdTemplate = "mpv --no-audio-display --pause %f --input-ipc-server=%s --audio-channels=stereo --audio-samplerate=48000 --audio-format=s16 --ao=pcm --ao-pcm-file=/tmp/snapfifo"

Jukebox.Default = "snap"
Jukebox.Enabled = true
```

**Reply by tcurdt** [reply]
I made some progress - but it seems like Navidrome fails to connect to MPV.

https://github.com/navidrome/navidrome/issues/3256#issuecomment-2571479292

**Reply by tcurdt** [reply]
I got it somewhat working...

- using jamstash as the subsonic client
- but progress updates are not working
- it seems play:sub needs the progress updates and without does not work at all
- the mpv template quoting situation is a bit of a mess

https://github.com/navidrome/navidrome/issues/3619
https://github.com/navidrome/website/pull/187

**Comment by ms140569** [comment]
Hi Torsten, here's Matthias Schmidt. I did the initial Jukebox-Mode implementation. I (initially) missed badly some sort of remoting or streaming to satellites/players in Navidrome. Something like the remoting included in the basic model of the Squeezebox Server for example. But Alas, the feature of connecting players or satellites to Navidrome is not included in the original (Subsonic) protocol/model. The more I think about this, the more I come to the believe that Navidrome should not reinvent the wheel and come up with yet another streaming mechanism. We should enable to use everything which is out in the field in the hope that some solutions are tested and stable and could be used without hazzle. But I totally agree with you that we could do way better to connect Navidrome to those streaming solutions (quoting, fork/exec, ...)

**Reply by tcurdt** [reply]
Thanks for the reply, Matthias. And thanks for working on the Jukebox mode. ❤️

The original idea and how I came to open the discussion was actually so simple - it would not even re-invent anything. It would still be the very same protocol. And all clients would work as is - that's the beauty. The satellites would just appear as different servers and proxy to the "root" server. This should make it a **very** low hanging fruit based on the existing jukebox support.

That said...

After looking into snapcast - it's hard to argue this doesn't come with some huge benefits. I got a Pi Zero running the snapcast client and at least running that snapcast client was darn easy. The fact that snapcast even supports multi-room audio is pretty hard to beat. So maybe those features beat the simplicity of the proxy satellites approach.

Either way...

There are a couple of things that we can hopefully improve - one way or another.

For one it feels like the current configuration could benefit from some rethinking. Both `Jukebox.Devices` and `MPVCmdTemplate` feel a bit problematic and seem to have outgrown their original use cases. So maybe we could come up with something that works better for all situations. For local and remote audio targets.

The fact that we decode and re-encode the data for the snapcast transmission is a little sad. But I am not sure we would gain that much by re-implementing native snapcast support. It would still require transcoding to the supported formats (opus, flac, vorbis). That said, it should be totally possible to decode to PCM and write it to a fifo from navidrome. What I would like to avoid if possible is sending PCM over the air. IIRC the original AirPlay did that.

The things I am currently worried about is
- the snapcast play progress support that seems broken (for me)
- general client support (I am using feishin/audioling and play:sub currently)
- the delay when starting/stopping

**Reply by tcurdt** [reply]
Any further thoughts on this, @ms140569

**Reply by tcurdt** [reply]
ping

**Reply by daredoes** [reply]
@tcurdt @ms140569 @deluan  I've been troubleshooting this a bit and by passing `--log-file=/data/mpv_debug.log` into the MPVCMDTemplate I've been able to figure out that the issue during playback with all of these workarounds is how the PCM file option seems to dump the entire song to a file instantly, instead of streaming it FIFO style. 

Here is my log

```log
[   0.066][i][ao/pcm] File: /tmp/navidrome (WAVE)
[   0.066][i][ao/pcm] PCM: Samplerate: 44100 Hz Channels: 2 Format: s16
[   0.066][v][ao/pcm] device buffer: 65536 samples.
[   0.066][v][ao/pcm] using soft-buffer of 65536 samples.
[   0.066][i][cplayer] AO: [pcm] 44100Hz stereo 2ch s16
[   0.066][v][cplayer] AO: Description: RAW PCM/WAVE file writer audio output
[   0.066][v][autoconvert] inserting resampler
[   0.066][v][swresample] format change, reinitializing resampler
[   0.066][v][swresample] 44100Hz stereo floatp -> 44100Hz stereo s16
[   0.066][v][af] [out] 44100Hz stereo 2ch s16
[   0.067][d][file] resize stream to 131072 bytes, drop 393216 bytes
[   0.068][v][cplayer] audio ready
[   0.068][v][cplayer] starting audio playback
[   0.068][v][cplayer] playback restart complete @ 0.000001, audio=playing, video=eof (paused)
[   0.068][v][cplayer] Set property: user-data/osc/margins={"t":0,"l":0,"r":0,"b":0} -> 1
[   0.085][d][osc] osc_init
[   0.086][v][cplayer] Set property: user-data/osc/margins={"t":0,"l":0,"r":0,"b":0} -> 1
[   0.091][v][ipc_0] Client connected
[   0.092][v][cplayer] Set property: pause=false -> 1
[   0.092][v][ao/pcm] starting AO
[   0.132][v][lavf] EOF reached.
[   0.134][v][af] filter input EOF
[   0.134][v][af] filter output EOF
[   0.134][v][cplayer] audio filter EOF
[   0.134][v][ao/pcm] audio end or underrun
[   0.134][v][cplayer] Set property: volume=100 -> 1
[   0.134][v][cplayer] AO signaled EOF (while in state playing)
[   0.134][v][cplayer] EOF code: 1  
```

The `lavf` demuxers and others reach the EOF before the `ao/pcm` which complains about `audio end or underrun`. Then the MPV command exits, and this cascades into closing the socket ruining the follow up jukebox command, but the tmp file has a whole song or playlist dumped into it and begins to play that. This happens whenever a "Track" is played.

My thoughts on a solution:
When jukebox mode is enabled the MPV socket should be established on server start, and `--keep-open` should be used to maintain its state. When the server is closed, the mpv socket should be closed, ideally leaving the tmp file empty at this point in time. When a track is played, the socket should be used to clear/add/append the playlist with the relevant data based on the action.

Gonna do some more testing/experimenting to see if that works locally. Would love any thoughts on if this makes sense to any of you. It seems like there isn't much of an alternative to using MPV, and based on the manpages for MPV this has nothing to do with the options being passed in, or the quoting of the strings, or whatever.

**Comment by mist2t** [comment]
I was just researching how to use Navidrome with "hardware" players / streamers like Wiim, Eversolo etc.

Implementing a feature where Navidrome can play directly to various endpoints / satellites thru the house I think it's essential.
Similar to what Roon does (or plexamp headless for that matter).

I think this also might be correlated with the "client" capabilities but maybe we can use something like raspberry py's ... or other solutions as "endpoints". This might be harder since Navidrome doesn't have official player software (only the web interface), so relying on third party it's a roadblock.

Too bad that hardware manufacturers don't implement the subsonic API natively. They supported Roon for some time and recently some of them started to add Plexamp support.

We need something for Navidrome. It's essential !
Meanwhile I'm thinking about using an iPad hooked up as a "hardware" music player, directly to DAC/AMP.

**Reply by michaelkrieger** [reply]
This is where navidrome acting as a sever decides the audio and passes it to Sonos, amplifiers, android-based-tvs, Apple TVs, fire sticks get data over airplay or chomecast. [mkchromecast](https://github.com/muammar/mkchromecast/wiki/FAQ) for example.

**Reply by tcurdt** [reply]
A work around for me for now:
I have installed music assistant which connects to navidrome and streams music via snapcast to my remotes.
Not a great solution as the UI is not great and it's yet another app - but at least it's something.

**Reply by GeorgeHahn** [reply]
I do wonder if Wiim would add Navidrome support if Navidrome had an appropriate protocol for the purpose. They have LMS support and I don't believe the OpenSubsonic protocol is suitable. I'm not sure if Navidrome would be better to adopt an existing remote client protocol (like LMS') or create its own and advocate for support.

**Comment by nis65** [comment]
My usecase is a bit different:

* I have access to a navidrome server with a named account somewhere in the internet, managed by someone else.
* At home, I have a snapcast multiroom setup based on raspberry pis with one snapserver and multiple snapclients.
* On one of the Raspberries, I even have connected my analog record player so that I can stream my old vinyl records into all rooms. 
* Another input to snapserver is the audio streamin output of an mpd. Having seen navidrome in action, I am thinking about replacing mpd with it.

What I would need is that I can access **the same user session** on the external navidrome **from two different devices*
* my laptop or mobile to **control** what is playing
* my raspberry pi to access the **music stream** and feed it to my snapserver.

Is this possible? does the subsonic API have the possibility to access the music stream that is controlled from a different device?

**Reply by tcurdt** [reply]
What would the local navidrome be doing? Are you maybe looking for something like music assistant?

**Reply by nis65** [reply]
No, my current solution is based on [this repo](https://github.com/Daenou/ansible-multiroom-audio) to configure my raspberry pi zoo which in turn is based on [this architecture](https://github.com/nis65/multiroom) (a bit outdated).

What I want to achieve is:

* I want to continue to use my local `mpd` (or, maybe in the future, `navidrome`) to play the music from my local collection as daily driver. This shall stay as is, I want to be able to listen to music in house even when the internet connection is dead.
* But from time to time, I would like to use the remote `navidrome` like a remote "radio station" to inject the music from there into my local multiroom system. So I would use my WebBrowser to connect to that remote navidrome instance to control the music to be played, **but** the music shall **not** be rendered by my local browser (on the loudspeakers of the laptop), but I would like raspberry Pi to fetch the music from my remote navidrome session and inject it into my local snapcast server.

This is why I was asking if I can "fetch" the music stream from an established navidrome session from a different device than the one that controls navidrome.

This is like "spotify connect" where I can choose to play the music not on the mobile I use to control what is playing, but choose a different "audio renderer" (or audio sink or loudspeaker). I actually implemented this (for spotify) in [this ansible role](https://github.com/Daenou/ansible-multiroom-audio/tree/main/roles/raspotify) reusing `raspotify`.

**Reply by tcurdt** [reply]
As I pointed out, music assistant gets you much closer to this as you can define separate music and player providers. You would just need to add your 2nd navidrome setup as music provider.

It's a little sad - but I gave up hope on the whole jukebox support for navidrome. And I am using music assistant to play my navidrome music in a multiroom setup. Unfortunately the MA control plane is still subpar and last time I checked there was only browser support. Multiuser support only just about to go into beta. But combining MA and navidrome is the best solution I found so far without me screaming into a pillow.

Say what you want about Spotify but their Spotify Connect sets the bar on how things should work from the UX perspective.
I still think should be possible to add something like this to navidrome - but it needs more people pushing for it.

**Reply by nis65** [reply]
I just tested *does the subsonic API have the possibility to access the music stream that is controlled from a different device* with the following setup:

* login to the remote navidrome with firefox, start playing a song
* login to the remote navidrome with chromium. And I don't see what the other session is playing, I even can play a different song (now hearing both at the same time on my laptop :scream:).

So the answer is possibly **no**, the subsonic API does not support (yet) what I want to achieve...

**Comment by ka-mp-6765** [comment]
I'd like to take it a step further and not only have satellites tell the current playing satellite what to play/skip song/etc. 
but .... satellites taking over from one another.  t.ex. when I start with an audiobook on my pc at work, I pause it at the end of the day and get into my car, press play on my phone and continue where it was paused. Then at home my tablet takes over from the phone etc etc etc

**Comment by mist2t** [comment]
There is a new FOSS whole-house, multi-room audio protocol out in the wild launched by the **Music Assistant** team, called "**Sendspin**".
 
It's really awesome. It's similar to Roon / Sonos and enables lossless and hi-res, synchronized, music playback from the server to various endpoints clients running the Sendspin client (there is already various SDKs / libraries for it).

Details: https://www.sendspin-audio.com/

I'd wish Navidrome will implement the Sendspin server. 
It's FOSS, backed by an amazing team (Music Assistant) and I envision down the line will earn support from various retail audio hardware (ex: Wiim) or some amazing little endpoints based on Raspberry PI's or ESP32 will emerge.

**Reply by tcurdt** [reply]
This sounds really cool.

But frankly, it's the clients that need to support controlling the server for this to work well enough.
Chicken-and-egg problem?

**Reply by mist2t** [reply]
It works pretty awesome already in Music Assistant.

From my understanding (not a dev though) the server (ex Navidrome backend) must implement the server side of Sendspin ... then all compatible Sendspin clients will receive the audio.

They advertise on the network using mDNS etc.

So, yeah, both have to implement it (server and clients).

I think if the server (Navidrome) will push the rock downhill by implementing it server side ... other clients will follow.

It's such a cool thing and i think it will pick up a lot of support down the line.

Open Home Foundation (Home Assistant and Music Assistant guys) are behind it and i think it will be here to stay, the de facto FOSS home audio protocol.

**Reply by tcurdt** [reply]
Well, I have it working with snapcast.
So, what are the benefits?

It would be interesting what clients you are using. For me, music assistant is only the glue to get remote speakers working currently. The MA web interface is not a serious solution - at least for me. And I barely use my remote speakers because of it.

I am currently using play:sub as a client on mobile and feishin on desktop. Unless there is enough interest to get this integrated into the common clients, at least I am not eager spending more time on this.

I don't think adding the server is much of a problem, the client support is.

**Reply by daredoes** [reply]
@tcurdt I'm right there with you on Snapcast via Music Assistant on my media server connected to my HA server over the network. What works for me is putting all my remote speakers in groups via Music Assistant (that checkbox sync thing). It only works well if all the items in the same group have the same source (snapcast, airplay, etc). I also have custom named IDs for the audio entities in music assistant, so if something breaks with an update or I change my snapcast server I just reset those IDs and my Home Assistant dashboard music functions start working again.

I've got spotify-connect hooked up via Music Assistant at this point, but it is horrid when trying to change volume. Might hop in as a dev to fix this myself.

Airplay via Music Assistant via Snapcast has been working pretty well though. 

Local support via Music Assistant has been good too.

**Reply by shrippen** [reply]
This would be a really great addition. But yes it would require a lot of effort accross the subsonic landscape because server and clients have to support it. But then again - the result would push it quite a lot further!

**Reply by tcurdt** [reply]
I still think this would need some refactoring to something like this

```
Jukebox.Devices = [
    [ "internal", "coreaudio/BuiltInSpeakerDevice" ],
    [ "speaker1", "sendspin-speaker1-address" ],
    [ "speaker2", "sendspin-speaker2-address" ]
]

Jukeboxes = [
    [ "default", "internal" ],
    [ "everwhere", "speaker1", "speaker2" ],
    [ "userA", "speaker1" ]
]
```

...but unless a committer is on board this discussion is pointless.

Sending to sendspin should become rather simple and native with https://github.com/Sendspin/sendspin-go
Although I do not yet see support for seeking yet.

But Jukebox client support still is the main problem.
It would be nice to have some iOS/Android app authors on board for that effort.

**Reply by shrippen** [reply]
I have mentioned this discussion to Tolriq on the Symfonium forum: https://support.symfonium.app/t/sendspin-support/13311

**Comment by hungryandgrumpy** [comment]
I’m interested in working on this, specifically starting with Google Cast as a first-class server-side playback target.

My current setup is Navidrome → Music Assistant → Google Cast speakers, and I’d like to remove the intermediary entirely.

The behaviour I’m aiming for is:

* select a Cast device from the Navidrome web UI
* start a track/album/queue
* Navidrome owns the playback session
* playback continues if the browser is closed
* reopening Navidrome from another browser shows the current target/playback state

I’m currently inspecting the existing Jukebox implementation to see whether this can be introduced as a general playback/output abstraction rather than as a Cast-specific parallel subsystem.

Tentatively I’m thinking of splitting the work into:

1) generalise the existing playback/output architecture while preserving current Jukebox behaviour
2) add a Google Cast provider and minimal target-selector UI
3) follow with optional UX improvements separately

I’d like to keep the changes upstream-friendly. Does that direction fit with how maintainers see Jukebox/remote playback evolving?

**Reply by tcurdt** [reply]
I would not focus on the web frontend. This needs to work with other clients (iOS/Android) as well. Otherwise this is a dead-end.

**Reply by hungryandgrumpy** [reply]
Agreed that this shouldn't be tied architecturally to the web frontend.

My intention is for playback and Cast control to be entirely server-side. The web UI would only be the first client of a target-selection/control API; closing it would have no effect on playback.

That should leave the door open for iOS/Android/Subsonic clients to use the same server-side functionality later.

For the initial work I'm looking at the existing Jukebox playback core first, independently of any frontend. The first change I'm considering is simply removing its hard dependency on MPV so other playback backends can be supported without changing existing Jukebox behaviour.

The question of how target discovery/selection should be exposed to non-web clients is important, but I'd prefer to keep that separate from the playback backend refactor rather than bake web-specific assumptions into it.

**Reply by tcurdt** [reply]
As said before, IMO we should aim at an abstraction like this

```
Jukebox.Speakers = [
    [ "internal", "coreaudio/BuiltInSpeakerDevice" ],
    [ "speaker1", "sendspin-speaker1-address" ],
    [ "speaker2", "sendspin-speaker2-address" ]
]

Jukeboxes = [
    [ "default", "internal" ],
    [ "everwhere", "speaker1", "speaker2" ],
    [ "userA", "speaker1" ]
]
```

And I personally would rather go for the open source `sendspin` direction, but (except for possible licensing questions) I don't see why there couldn't also be a Chrome Cast speaker.

But this also comes down to clients supporting Jukebox mode properly.

**Reply by nis65** [reply]
For me, quality is more important than minimal latency. So I would prefer a solution that can transport FLAC from source to the speakers with network bandwidth being the only limit. I work a lot with snapcast and like it very much.

Is there a fundamental advantage of sendspin over snapcast?

**Reply by tcurdt** [reply]
`sendspin` seems to be easier to integrate when I looked at it. But frankly we could have any number of speaker implementations.
But since we are already talking about three now, it makes me wonder if they should be part of `navidrome`.
All need to be maintained.

**Reply by nis65** [reply]
> it makes me wonder if they should be part of `navidrome`

I also think from an architecture perspective it would make more sense to just have a generic audio sink (eg. `alsa`) where other solutions can get the audio data from. However, as the goal is to provide each individual user with its own audio stream, it could make more sense to look into icecast or sth similar.

My original requirement was kind of:
* I login with some sort of client (e.g. web) and want to control what is playing.
* i want a different client (from a different system) to have access to that audio stream. I think the wire protocol should be still `http` so that we can make use of authentication etc.  And I want to be able to render that audiostream not only to a real sound card, but also e.g. to a snapserver.

Simply put: for me it would be enough if I can login to `navidrome` from two different browsers while browser 1 controls what is playing and browser 2 gets the audiostream controlled by browser 1. Handing over the audio from the browser to snapserver can be solved outside of `navidrome`.

**Reply by tcurdt** [reply]
The audio sink is where I very much disagree. The whole point of Chrome Cast an friends is to NOT send raw audio and to NOT re-encode. Which is why an audio sink is is very much the opposite of what I have in mind here.

For me the whole idea is to tell the client where to find the audio data in whatever format - not send raw audio.

**Reply by nis65** [reply]
@tcurdt You are right, the audio sink is conceptionally wrong. Thanks for interfering.

What about the idea of just a small protocol enhancement of the current protocol that would allow me to say when login "do not create a new session, just deliver the audio to me instead of the first session".

I understand that would break the possibility to run multiple independent sessions with the same user account, but wouldn't that be a first start? Do we really need a new protocol to distribute the audio data?

**Reply by tcurdt** [reply]
I am not sure what protocol enhancement you have in mind - clients would need to understand this.

Question is whether the existing Jukebox mode could/should be sufficient. Not so sure about a speaker selection.

Which is why I keep saying we need client authors on board with this.

I think it would be good to have/use an open source client with Jukebox mode support and then trying to implement this in tandem.

**Reply by hungryandgrumpy** [reply]
Small update from the experiment I mentioned earlier: this is now working on real hardware.

I have a native server-side Google Cast implementation running through Jukebox. Navidrome owns the playback/session, and the Cast receiver fetches the media directly from Navidrome rather than Navidrome sending decoded audio to it.

On my Harman Kardon Cast group I now have:

* server-owned playback
* play/pause/seek
* Navidrome → speaker volume control
* speaker hardware volume buttons → Navidrome Jukebox gain

That last bit was particularly rewarding to see working.

The implementation has been race-tested and hardware-tested, including rebuilding and testing the final cleaned tree without the diagnostic experiments we used while debugging.

One useful thing the hardware test exposed: selecting the correct Cast target is important. The working target for my setup is the logical Cast group endpoint, not simply one speaker’s normal :8009 endpoint. So discovery/stable target identity now looks like a real correctness requirement, not just UI polish.

I’ve opened a draft PR here: [https://github.com/navidrome/navidrome/pull/5946](https://github.com/navidrome/navidrome/pull/5946   )

I’ve deliberately kept discovery out of this PR so the Cast transport/playback part stays reviewable.

**Reply by tcurdt** [reply]
Nice. The biggest problem for me was always seek not working properly.
How do you select the speaker/group or is it hardwired?

**Reply by hungryandgrumpy** [reply]
@tcurdt , it is no longer hardwired. 

Draft PR [https://github.com/hungryandgrumpy/navidrome/pull/1 ](https://github.com/hungryandgrumpy/navidrome/pull/1 )adds discovery of Cast receivers/groups and lets Navidrome select which target the server-side Jukebox controls.

Each target keeps its own queue/session, so changing the selected target does not migrate or stop another target’s existing playback.

A dedicated picker UI is still pending — this draft is the server-side discovery/selection/API plumbing only.

On seek: it is wired through the Cast playback path, but I haven’t done a dedicated hardware acceptance pass for seek yet, so I don’t want to overclaim that part.
